pkg/kube/fakeclient.go

```diff
@@ -0,0 +1,397 @@
+		return func(action clienttesting.Action) (handled bool, ret watch.Interface, err error) {
+			gvr := action.GetResource()
+			ns := action.GetNamespace()
+			watch, err := tracker.Watch(gvr, ns)
+			if err != nil {
+				return false, nil, err
+			}
+			c.informerWatchesPending.Dec()
+			return true, watch, nil
+		}
+	}
+	for _, fc := range []fakeClient{
+		c.kube.(*fake.Clientset),
+		c.istio.(*istiofake.Clientset),
+		c.gatewayapi.(*gatewayapifake.Clientset),
+		c.dynamic.(*dynamicfake.FakeDynamicClient),
+		c.metadata.(*metadatafake.FakeMetadataClient),
+	} {
+		fc.PrependWatchReactor("*", watchReactor(fc.Tracker()))
+	}
+	df.PrependReactor("list", "*", listReactor)
+	c.metadata.(*metadatafake.FakeMetadataClient).PrependReactor("list", "*", listReactor)
+
+	c.fastSync = true
+
+	c.version = lazy.NewWithRetry(c.kube.Discovery().ServerVersion)
+
+	if NewCrdWatcher != nil {
+		c.crdWatcher = NewCrdWatcher(c)
+	}
+
+	return c
+}
+
+func NewFakeClientWithVersion(minor string, objects ...runtime.Object) CLIClient {
+	c := NewFakeClient(objects...).(*client)
+	c.Kube().Discovery().(*fakediscovery.FakeDiscovery).FakedServerVersion = &kubeVersion.Info{Major: "1", Minor: minor, GitVersion: fmt.Sprintf("v1.%v.0", minor)}
+	return c
+}
+
+type fakeClient interface {
+	PrependReactor(verb, resource string, reaction clienttesting.ReactionFunc)
+	PrependWatchReactor(resource string, reaction clienttesting.WatchReactionFunc)
+	Tracker() clienttesting.ObjectTracker
+}
+
+func insertPatchReactor(f clienttesting.FakeClient, fmf *fieldManagerFactory) {
+	f.PrependReactor(
+		"patch",
+		"*",
+		func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+			pa := action.(clienttesting.PatchAction)
+			if pa.GetPatchType() == types.ApplyPatchType {
+				// Apply patches are supposed to upsert, but fake client fails if the object doesn't exist,
+				// if an apply patch occurs for a deployment that doesn't yet exist, create it.
+				// However, we already hold the fakeclient lock, so we can't use the front door.
+				// rfunc := clienttesting.ObjectReaction(f.Tracker())
+				original, err := f.Tracker().Get(pa.GetResource(), pa.GetNamespace(), pa.GetName())
+				var isNew bool
+				mygvk := gvk.MustFromGVR(pa.GetResource()).Kubernetes()
+				if kerrors.IsNotFound(err) || original == nil {
+					isNew = true
+					original = kubeclient.GVRToObject(pa.GetResource())
+					d := original.(metav1.Object)
+					d.SetName(pa.GetName())
+					d.SetNamespace(pa.GetNamespace())
+					e := original.(schema.ObjectKind)
+					e.SetGroupVersionKind(mygvk)
+				}
+
+				applyObj := &unstructured.Unstructured{Object: map[string]interface{}{}}
+				if err := json.Unmarshal(pa.GetPatch(), &applyObj.Object); err != nil {
+					log.Errorf("error decoding YAML: %v", err)
+					return false, nil, err
+				}
+				// TODO: I'm not 100% sure this should always force, but we lose the apply option...
```

Comment:
I bet we could add this upstream. Seems useful to have..

Reply:
Thanks, I like that idea!